### PR TITLE
[v6r16] MessageQueue resources

### PR DESCRIPTION
**Pull request by atsareg**
This PR contains classes for managing MQ connections as DIRAC Resources. The current implementation is only for the Stomp MQ protocol which is provided by the RabbitMQ. This is a wrapper over the MQConnection classes by Sergio as in #2974, and replaces this PR.

Users are supposed to use MQListener and MQPublisher classes which can be used as simply as the following:
  
publisher = MQPublisher( "MyQueue" )
listener = MQListener( "MyQueue" )
publisher.put( message )
listener.get( message )

Or:
publisher = MQPublisher( "MyQueue" )
listener = MQListener( "MyQueue", callback = myCallback )
publisher.put( message )
listener.run()

Example of the MQ configuration description is the following:

Resources
{
  MQServices
  {
    mardirac3.in2p3.fr
    {
      MQType = Stomp
      Host = mardirac3.in2p3.fr
      Port = 9165
      User = guest
      Password = guest
      Queues
      {
        TestQueue
        {
          Acknowledgement = True
        }
      }
    }
  }
}

The queueName in the listener/publisher instantiation can be given as just the Queue name (if no ambiguities ) or fully qualified name like "mardirac3.in2p3.fr::TestQueue" to avoid ambiguities.

### CODE REVIEW

File: Resources/MessageQueue/MQConnection.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    :param dict parameters: dictionary with the parameters for the queue. It should include the following parameters:
+    'Host', 'Port', 'User', 'VH' and 'Type'. Otherwise, the function will return an error
+    """
+
+    if not parameters:
+      return S_ERROR( EMQUKN, 'Queue parameters are not provided' )
+
+    for param in [ 'Host', 'Port', 'User', 'VHost' and 'Type' ]:
+      if not param in parameters:
```
Comment (andresailer) [inline Resources/MessageQueue/MQConnection.py]: `param not in parameters` is preferred

File: Resources/MessageQueue/MQConnection.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    """ Sets the parameters attribute
+
+    :param dict parameters: dictionary with the parameters for the queue. It should include the following parameters:
+    'Host', 'Port', 'User', 'VH' and 'Type'. Otherwise, the function will return an error
+    """
+
+    if not parameters:
+      return S_ERROR( EMQUKN, 'Queue parameters are not provided' )
+
+    for param in [ 'Host', 'Port', 'User', 'VHost' and 'Type' ]:
```
Comment (andresailer) [inline Resources/MessageQueue/MQConnection.py]: The `and` should be a comma, right?

File: Resources/MessageQueue/MQConnection.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    :param bool receive: flag to enable the MQ connection for getting message
+    :param func messageCallback: function to be called when a new message is received from the queue ( only receiver mode ).
+                                If None, the defaultCallback method is used instead
+    :return: S_OK/S_ERROR
+    """
+    raise NotImplementedError( 'This method should be implemented by child class' )
+
+  def run( self, parameters = {}, receive = True, messageCallback = None ):
+    """
+    Establishes a new blocking connection to the message queue
```
Comment (andresailer) [inline Resources/MessageQueue/MQConnection.py]: need an empty line before `:param...`

File: Resources/MessageQueue/MQConnection.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+import time
+
+from DIRAC import gLogger, S_OK, S_ERROR
+from DIRAC.Core.Utilities.DErrno import EMQUKN
+
+__RCSID__ = "$Id$"
+
+class MQConnectionError( Exception ):
+    def __init__(self,*args,**kwargs):
```
Comment (chaen) [inline Resources/MessageQueue/MQConnection.py]: Do you actually need to implement the constructor if it is just forwarding the parameters ?

File: Resources/MessageQueue/MQConnection.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    :param dict parameters: dictionary with the parameters for the queue. It should include the following parameters:
+    'Host', 'Port', 'User', 'VH' and 'MQType'. Otherwise, the function will return an error
+    """
+
+    if not parameters:
+      return S_ERROR( EMQUKN, 'Queue parameters are not provided' )
+
+    for param in self.MANDATORY_PARAMETERS:
+      if param not in parameters:
+        return S_ERROR( EMQUKN, "Parameter %s not provided" % param )
```
Comment (chaen) [inline Resources/MessageQueue/MQConnection.py]: You could make a set difference to return the list of missing parameters ? So in one call we know all the missing params

File: Resources/MessageQueue/MQConnection.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    self.parameters = parameters
+
+    return S_OK( 'Queue parameters set successfully' )
+
+  def setupConnection( self, parameters = {}, receive = False, messageCallback = None ):
```
Comment (chaen) [inline Resources/MessageQueue/MQConnection.py]: ```{}``` is a dangerous default argument, it should rather be ```None``` and isntanciated in the method if needed

File: Resources/MessageQueue/MQConnection.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    :param dict parameters: dictionary with additional MQ parameters if any
+    :param bool receive: flag to enable the MQ connection for getting message
+    :param func messageCallback: function to be called when a new message is received from the queue ( only receiver mode ).
+                                If None, the defaultCallback method is used instead
+    :return: S_OK/S_ERROR
+    """
+    raise NotImplementedError( 'This method should be implemented by child class' )
+
+  def run( self, parameters = {}, receive = True, messageCallback = None ):
```
Comment (chaen) [inline Resources/MessageQueue/MQConnection.py]: same here for ```{}``` as default

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        return result
+      queueParameters = result['Value']
+    if parameters:
+      queueParameters.update( parameters )
+
+    mqType = queueParameters.get( 'MQType' )
+    if not mqType:
+      mqType = self.mqType
+    if not mqType:
+      return S_ERROR( 'No MQType specified' )
```
Comment (chaen) [inline Resources/MessageQueue/MQConnectionFactory.py]: Error number ?

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    ceClass = result['Value']
+    try:
+      mqConnection = ceClass()
+      result = mqConnection.setParameters( queueParameters )
+      if not result['OK']:
+        return result
+
+    except Exception as x:
+      msg = 'MQConnectionFactory could not instantiate %s object: %s' % ( subClassName, str( x ) )
+      self.log.exception()
```
Comment (chaen) [inline Resources/MessageQueue/MQConnectionFactory.py]: aren't you missing the parameters of this log call ?

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      self.log.warn( msg )
+      return S_ERROR( msg )
+
+    return S_OK( mqConnection )
+
+  def getMQListener( self, queueName = None, parameters = {} ):
```
Comment (chaen) [inline Resources/MessageQueue/MQConnectionFactory.py]: same here for ```{}``` as default

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    """
+    result = self.__getMQConnection( queueName = queueName,
+                                     parameters = parameters )
+    if not result['OK']:
+      return result
+
+    mqConnection = result['Value']
+    return S_OK( mqConnection )
+
+  def getMQPublisher( self, queueName = None, parameters = {} ):
```
Comment (chaen) [inline Resources/MessageQueue/MQConnectionFactory.py]: same here for ```{}``` as default

File: Resources/MessageQueue/StompMQConnection.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    self.receiver = False
+    self.acknowledgement = False
+    self.subscriptionID = None
+    self.msgList = []
+    self.lock = threading.Lock()
+    self.connection = None
+
+  # Callback functions for message receiving mode
+
+  def on_message( self, headers, message ):
```
Comment (chaen) [inline Resources/MessageQueue/StompMQConnection.py]: By whom is this called ? directly by the stomp library ?

File: Resources/MessageQueue/test/Test_MQ_test.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+  def setUp(self):
+
+    mqCFG = CFG()
+    mqCFG.loadFromBuffer( TEST_CONFIG )
+    gConfigurationData.mergeWithLocal( mqCFG )
+    mqCFG.loadFromBuffer( TEST_BAD_CONFIG )
+    gConfigurationData.mergeWithLocal( mqCFG )
+
+  @unittest.skipIf( not checkHost( 'mardirac3.in2p3.fr', 9165 ), "Test MQ host is not reachable" )
```
Comment (chaen) [inline Resources/MessageQueue/test/Test_MQ_test.py]: Could you maybe just put these as global parameters in the head of the file ?

**Comment by chaen** [comment]
Just to make sure: the names of the MQServices do not need to be host names right ? 

Also, I am a bit lost with all these PRs, is there already some docs for all these ?

**Comment by atsareg** [comment]
Names of the MQServices do not need to be host names.

We are just establishing now the MQ architecture, not yet documented. This will come of course

**Comment by fstagni** [comment]
Administrator docs and developer docs should be included. What you've put in the first message when opening the PR can certainly be part of it.

File: Resources/MessageQueue/MQConnection.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    """
+
+    if not parameters:
+      return S_ERROR( EMQUKN, 'Queue parameters are not provided' )
+
+    for param in self.MANDATORY_PARAMETERS:
+      if param not in parameters:
+        return S_ERROR( EMQUKN, "Parameter %s not provided" % param )
+
+    self.parameters = parameters
```
Comment (fstagni) [inline Resources/MessageQueue/MQConnection.py]: Define it in `__init__`

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+    ceClass = result['Value']
+    try:
+      mqConnection = ceClass()
+      result = mqConnection.setParameters( queueParameters )
+      if not result['OK']:
+        return result
+
+    except Exception as exc:
+      msg = 'MQConnectionFactory could not instantiate %s object: %s' % ( subClassName, str( exc ) )
```
Comment (fstagni) [inline Resources/MessageQueue/MQConnectionFactory.py]: Better use `repr(exc)`. And anyway, which exception can be thrown here?

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+      self.log.exception( 'Could not instantiate MQConnection object', IException = exc )
+      self.log.warn( msg )
+      return S_ERROR( msg )
```
Comment (fstagni) [inline Resources/MessageQueue/MQConnectionFactory.py]: New structure?

File: Resources/MessageQueue/MQConnectionFactory.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+    return S_OK( mqConnection )
+
+  def getMQListener( self, queueName = None, parameters = None ):
```
Comment (fstagni) [inline Resources/MessageQueue/MQConnectionFactory.py]: I think this method can be just coded
```
return __getMQConnection( queueName = queueName,  parameters = parameters )
```

File: Resources/MessageQueue/MQConnection.py [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+
+  MANDATORY_PARAMETERS = [ 'Host', 'Port', 'User', 'VHost', 'MQType' ]
+
+  def __init__( self ):
+    """ Standard constructor
+    """
+
+    self.alive = False
+
+  def on_error( self, headers, message ):
```
Comment (zmathe) [inline Resources/MessageQueue/MQConnection.py]: headers is not used you can remove it

File: Resources/MessageQueue/StompMQConnection.py [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+    # Make the actual connection
+    host = self.parameters.get( 'Host' )
+    port = self.parameters.get( 'Port', 61613 )
+    vhost = self.parameters.get( 'VHost' )
+    user = self.parameters.get( 'User', 'guest' )
+    password = self.parameters.get( 'Password', 'guest' )
+    self.queueName = self.parameters.get( 'Queue' )
+    headers = {}
+    if "Persistent" in self.parameters and self.parameters['Persistent'].lower() in ['true', 'yes', '1']:
```
Comment (JanEbbing) [inline Resources/MessageQueue/StompMQConnection.py]: Also use `if self.parameters.get( 'Persistent', '' ).lower() in [ 'true', 'yes', '1' ]:` here?

File: Resources/MessageQueue/StompMQConnection.py [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+  def put( self, message ):
+    """
+    Sends a message to the queue
+    message contains the body of the message
+
+    :param str message: string or any json encodable structure
+    """
+    try:
+      if isinstance( message, list ):
```
Comment (JanEbbing) [inline Resources/MessageQueue/StompMQConnection.py]: replace this with 
`if not isinstance( message, list ):
  message = [ message ]`
remove else branch, execute for loop unconditionally?

File: Resources/MessageQueue/MQConnectionFactory.py (hunk 15 above)
Comment (atsareg) [inline Resources/MessageQueue/MQConnectionFactory.py]: This is a call to a third party library. God knows what can happen there

File: Resources/MessageQueue/MQConnection.py (hunk 18 above)
Comment (atsareg) [inline Resources/MessageQueue/MQConnection.py]: I guess this is a standard call signature for callbacks

File: Resources/MessageQueue/StompMQConnection.py (hunk 20 above)
Comment (atsareg) [inline Resources/MessageQueue/StompMQConnection.py]: This modifies input argument which should be always avoided

**Pull request closed without merging**